pkg/agent/openflow/client_test.go

```diff
@@ -152,58 +150,33 @@ func TestConcurrentFlowInstallation(t *testing.T) {
 			client := ofClient.(*client)
 			client.flowOperations = m
 
-			var wg sync.WaitGroup
-			var counter int32        // tracks the number of concurrent Add calls at any given time
-			var callIdx int32        // used to distinguish between the first Add call and subsequent calls
-			var concurrentCalls bool // set to true if we observe concurrent calls
-			var stop bool
-			var mutex sync.Mutex
-			cond := sync.NewCond(&mutex)
+			var concurrentCalls atomic.Bool // set to true if we observe concurrent calls
 
+			addCh := make(chan int)
 			m.EXPECT().Add(gomock.Any()).DoAndReturn(func(args ...interface{}) error {
-				cond.L.Lock()
-				defer cond.L.Unlock()
-				counter += 1
-				callIdx += 1
-				if callIdx == 1 { // first call
-					// we wait until we need to stop
-					// we stop when one of these conditions is met:
-					// 1) we detected concurrent Add calls (test is a success), or
-					// 2) the test timed out
-					for !stop {
-						cond.Wait()
-					}
-				} else if counter > 1 {
-					concurrentCalls = true
-					stop = true
-					cond.Signal()
+				x := rand.Int()
+				select {
+				case d := <-addCh:
+					assert.NotEqual(t, x, d)
+					concurrentCalls.Store(true)
+				case addCh <- x:
 				}
 				return nil
 			}).AnyTimes()
 
-			wg.Add(2)
-			go func() {
-				tc.installFn(ofClient, tc.cacheKey)
-				wg.Done()
-			}()
-			go func() {
-				tc.installFn(ofClient, tc.cacheKey)
-				wg.Done()
-			}()
-
-			// enforce a timeout of one second for the test
-			ctx, cancel := context.WithTimeout(context.Background(), time.Second)
-			go func() {
-				<-ctx.Done()
-				cond.L.Lock()
-				defer cond.L.Unlock()
-				stop = true
-			}()
+			var wg sync.WaitGroup
+			for i := 0; i < 2; i++ {
+				cacheKey := fmt.Sprintf(tc.cacheKeyFormat, i)
+				wg.Add(1)
+				go func() {
+					defer wg.Done()
```

Comment:
If someone is making changes to the OF `Client` implementation and for some reason concurrent calls are not possible any more, then the test will hang. Should we find a way to add a reasonable timeout for the test (like the one I had before).

For example, when I pull in your changes and use the following patch (just for testing purposes), I get into a deadlock situation, which is not detected by go:
```
diff --git a/pkg/agent/openflow/client.go b/pkg/agent/openflow/client.go
index 4a1e925..6e67c2b 100644
--- a/pkg/agent/openflow/client.go
+++ b/pkg/agent/openflow/client.go
@@ -98,6 +98,8 @@ func (c *client) GetFlowTableStatus() []binding.TableStatus {
 // added to the flow cache. If the flow cache has not been initialized yet (i.e. there is no
 // flowCacheKey key in the cache map), we create it first.
 func (c *client) addMissingFlows(cache *flowCategoryCache, flowCacheKey string, flows []binding.Flow) error {
+       cache.Lock()
+       defer cache.Unlock()
        // initialize flow cache if needed
        fCacheI, _ := cache.LoadOrStore(flowCacheKey, flowCache{})
        fCache := fCacheI.(flowCache)
diff --git a/pkg/agent/openflow/pipeline.go b/pkg/agent/openflow/pipeline.go
index 3bfd095..e4df7c2 100644
--- a/pkg/agent/openflow/pipeline.go
+++ b/pkg/agent/openflow/pipeline.go
@@ -104,6 +104,7 @@ type FlowOperations interface {
 type flowCache map[string]binding.Flow
 
 type flowCategoryCache struct {
+       sync.Mutex
        sync.Map
 }
 
```

Reply:
Sorry I missed this, just added it.